### PR TITLE
refactor(versioning/distro-info): for future use in debian and ubuntu versioning

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
No behavior change

## Context

related to #14601 and blocks #14774 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

File: lib/modules/versioning/distro.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import dataFiles, { DataFile } from '../../data-files.generated';
+
+export type DistroInfoRecord = Record<string, string>;
+
+export default class DistroInfo {
+  private readonly _codenameToVersion = new Map<string, string>();
+  private readonly _distroInfo: DistroInfoRecord;
+
+  constructor(distroJsonKey: string) {
```
Comment (viceice) [inline lib/modules/versioning/distro.ts:9]: define a string type with allowed values.

**Review by JamieMagee — DISMISSED** [review]
Approved pending @viceice's changes

**Review by viceice — CHANGES_REQUESTED** [review]
still missing change

**Comment by Gabriel-Ladzaretti** [comment]
@viceice 
currently switching to fetch the codename data from:
https://github.com/renovatebot/renovate/pull/14881#issuecomment-1086751248

^^ instead of the current `distro-info` package

File: lib/modules/versioning/distro.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  eol_elts?: string;
+}
+
+type DistroDataFile = 'data/ubuntu-distro-info.json';
+
+export type DistroInfoRecord = Record<string, DistroSchedule>;
+
+export type DistroInfoRecordWithVersion = { version: string } & DistroSchedule;
+
+export default class DistroInfo {
```
Comment (viceice) [inline lib/modules/versioning/distro.ts:21]: please no default export

File: lib/modules/versioning/ubuntu/index.ts [hunk 3]
```diff
@@ -1,19 +1,21 @@
...
+export const urls = [
+  'https://changelogs.ubuntu.com/meta-release',
+  "https://debian.pages.debian.net/distro-info-data/ubuntu.csv'",
+];
 export const supportsRanges = false;
 
 // #12509
 const temporarilyUnstable = ['22.04'];
 
+const ubuntuJsonKey = 'data/ubuntu-distro-info.json';
```
Comment (viceice) [inline lib/modules/versioning/ubuntu/index.ts:16]: move to distro

File: tools/distro-json-generate.mjs [hunk 4]
```diff
@@ -1,40 +1,69 @@
...
 
-if (!shell.which(`distro-info`)) {
-  shell.echo('This script requires distro-info, exiting...');
-  shell.exit(2);
+/**
+ * @param {string} url
+ * @param {function} cb
+ */
+function csv(url, cb) {
+  https.get(url, (response) => {
```
Comment (viceice) [inline tools/distro-json-generate.mjs:12]: use `got`, so we have simple promise api

File: tools/distro-json-generate.mjs [hunk 5]
```diff
@@ -1,40 +1,69 @@
...
-function objectify(str) {
-  let obj = {};
+  lines
+    // drop empty lines
+    .filter((l) => !!l)
+    .map((l) => {
+      const obj = {};
+      const line = l.split(',');
+      // eslint-disable-next-line
+      // @ts-ignore
```
Comment (viceice) [inline tools/distro-json-generate.mjs:53]: no, use jsdoc comments to declare types.

File: tools/distro-json-generate.mjs [hunk 6]
```diff
@@ -1,40 +1,69 @@
...
+      const ver = line.shift().replace(/LTS|\s/g, '');
 
-  for (const line of str.split(/\r?\n/)) {
-    let [ver, codename] = line.split(' ');
-    // eslint-disable-next-line
-    // @ts-ignore
-    obj[ver] = codename;
-  }
+      headers.map((h, i) => {
```
Comment (viceice) [inline tools/distro-json-generate.mjs:56]: use `for...of` loop

File: tools/distro-json-generate.mjs [hunk 7]
```diff
@@ -46,28 +75,36 @@ async function updateJsonFile(file, newData) {
 
   try {
     oldData = fs.existsSync(file) ? await fs.readFile(file, 'utf8') : null;
-    // Eliminate formatting
-    oldData = oldData?.replace(/\s/g, '') ?? null;
+    // Eliminate formatting. removes WS in the beginning, end. before & after non characters
+    oldData = oldData?.replace(/^\s|\s$|\B\s|\s\B/g, '') ?? null;
```
Comment (viceice) [inline tools/distro-json-generate.mjs:79]: why do we need this here? can we have those not added?

File: tools/distro-json-generate.mjs [hunk 8]
```diff
@@ -1,40 +1,69 @@
...
-/**
- * @param {string} str
- * @returns {{}}
- */
-function objectify(str) {
-  let obj = {};
+  lines
+    // drop empty lines
+    .filter((l) => !!l)
+    .map((l) => {
```
Comment (viceice) [inline tools/distro-json-generate.mjs:49]: use `for...of` loop. only use map when returned value is used.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/versioning/ubuntu/index.ts (hunk 3 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/index.ts:16]: I already have a string type in `distro` with this key.
other than that, how would u suggest instantiating `DistroInto` than ?

File: tools/distro-json-generate.mjs (hunk 7 above)
Comment (Gabriel-Ladzaretti) [reply tools/distro-json-generate.mjs:79]: used only to remove formatting before comparing old and new data and prevent writing to disk the same data every run

File: tools/distro-json-generate.mjs (hunk 5 above)
Comment (Gabriel-Ladzaretti) [reply tools/distro-json-generate.mjs:53]: switched to ts.

File: lib/modules/versioning/ubuntu/index.ts (hunk 3 above)
Comment (viceice) [reply lib/modules/versioning/ubuntu/index.ts:16]: moving those const to distro and create the combined type of this const.

File: package.json [hunk 9]
```diff
@@ -20,6 +20,7 @@
     "eslint": "eslint . --report-unused-disable-directives",
     "eslint-fix": "eslint --fix . --report-unused-disable-directives",
     "generate": "run-s generate:*",
+    "generate:distro-json-generate": "npx ts-node tools/distro-json-generate.ts",
```
Comment (viceice) [inline package.json:23]:
```suggestion
    "generate:distro-json-generate": "ts-node tools/distro-json-generate.ts",
```
we've ts-node as dev dep. 

we probably should add the alternative typescript Compiler for ts-node to speedup the execution and convert all tools to TS again for better type safety.

so please do the ts conversation in a separate PR.

File: tools/distro-json-generate.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
+import fs from 'fs-extra';
+import got from 'got';
+import shell from 'shelljs';
+
+const url = 'https://debian.pages.debian.net/distro-info-data/ubuntu.csv';
+
+/**
+ * @param {string} raw
+ * @returns {string}
+ */
```
Comment (viceice) [inline tools/distro-json-generate.ts:10]: not needed for TS. if used please extend to full description, so it's clear what the function is for.

File: package.json [hunk 11]
```diff
@@ -20,6 +20,7 @@
     "eslint": "eslint . --report-unused-disable-directives",
     "eslint-fix": "eslint --fix . --report-unused-disable-directives",
     "generate": "run-s generate:*",
+    "generate:distro-json-generate": "node tools/distro-json-generate.mjs",
```
Comment (viceice) [inline package.json:23]: @rarkins I don't think we should generate this file on every `generate` call. So this should be reverted. WDYT?

File: tools/distro-json-generate.mjs [hunk 12]
```diff
@@ -1,73 +1,105 @@
...
+      obj[h.replace('-', '_')] = line[i];
+    }
+
+    if (ver) {
+      // eslint-disable-next-line
+      // @ts-ignore
+      res[ver] = obj;
+    }
+  }
+  return JSON.stringify(res);
```
Comment (viceice) [inline tools/distro-json-generate.mjs:44]:
```suggestion
  return JSON.stringify(res, undefined, 2);
```
So file changes can be easier reviewed.

File: lib/modules/versioning/ubuntu/index.ts [hunk 13]
```diff
@@ -1,18 +1,16 @@
...
-  getVersionByCodename,
-  isCodename,
-} from './distribution';
 
 export const id = 'ubuntu';
 export const displayName = 'Ubuntu';
-export const urls = ['https://changelogs.ubuntu.com/meta-release'];
+export const urls = [
+  'https://changelogs.ubuntu.com/meta-release',
+  "https://debian.pages.debian.net/distro-info-data/ubuntu.csv'",
```
Comment (viceice) [inline lib/modules/versioning/ubuntu/index.ts:9]:
```suggestion
  'https://debian.pages.debian.net/distro-info-data/ubuntu.csv',
```

File: package.json (hunk 11 above)
Comment (viceice) [reply package.json:23]: @Gabriel-Ladzaretti Please revert this, we will update file on demand.

File: tools/distro-json-generate.mjs (hunk 5 above)
Comment (viceice) [reply tools/distro-json-generate.mjs:34]: Still an an issue

File: tools/distro-json-generate.mjs [hunk 14]
```diff
@@ -1,73 +1,106 @@
...
 
+  const oldLen = oldData?.length ?? -1;
+  const newLen = newAdj?.length ?? -1;
+
+  if (oldLen === -1 || newLen === -1 || oldLen > newLen) {
+    shell.echo(`New data might be corrupted!`);
+    shell.echo(`Aborting ${file} update`);
+    shell.echo(`**************** NEW DATA ****************\n${newData} `);
+    return;
+  }
```
Comment (viceice) [inline tools/distro-json-generate.mjs:82]: I think we don't need this here, should be covered by normal unit tests.

**Review by viceice — CHANGES_REQUESTED** [review]

File: tools/distro-json-generate.mjs (hunk 5 above)
Comment (Gabriel-Ladzaretti) [reply tools/distro-json-generate.mjs]: ah, finally got it :|. 
had a miss consumption about `jsdoc`, thought of if as sort of  `javadoc` used to comment function prototypes only. 

fixed.

File: tools/distro-json-generate.mjs (hunk 7 above)
Comment (Gabriel-Ladzaretti) [reply tools/distro-json-generate.mjs]: dropped it all together
as per:
> I think we don't need this here, should be covered by normal unit tests.

File: package.json (hunk 11 above)
Comment (Gabriel-Ladzaretti) [reply package.json]: reverted

File: lib/modules/versioning/ubuntu/index.ts (hunk 13 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/versioning/ubuntu/index.ts]: done

File: tools/distro-json-generate.mjs (hunk 14 above)
Comment (Gabriel-Ladzaretti) [reply tools/distro-json-generate.mjs]: cool, thats right. never thought of it this way, 
removed.

File: lib/modules/versioning/ubuntu/index.ts (hunk 13 above)
Comment (viceice) [reply lib/modules/versioning/ubuntu/index.ts]: please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#resolve-review-comments-instead-of-commenting

File: lib/modules/versioning/ubuntu/index.ts [hunk 15]
```diff
@@ -39,20 +37,23 @@ function isSingleVersion(version: string): boolean {
...
 function isStable(version: string): boolean {
-  const ver = getVersionByCodename(version);
+  const ver = di.getVersionByCodename(version);
   if (!isValid(ver)) {
     return false;
   }
-  if (temporarilyUnstable.includes(ver)) {
+
+  const schedule = di.getSchedule(ver);
+  if (schedule && new Date(schedule.release) > new Date()) {
```
Comment (viceice) [inline lib/modules/versioning/ubuntu/index.ts:46]: please use DateTime from luxon.

File: tools/distro-json-generate.mjs [hunk 16]
```diff
@@ -1,73 +1,77 @@
...
+      continue;
+    }
 
-  for (const line of str.split(/\r?\n/)) {
-    let [ver, codename] = line.split(' ');
-    // eslint-disable-next-line
-    // @ts-ignore
-    obj[ver] = codename;
-  }
+    /** @type {{[index: string]:any}} */
```
Comment (viceice) [inline tools/distro-json-generate.mjs:30]:
```suggestion
    /** @type {Record<string, any>} */
```

**Review by viceice — APPROVED** [review]
